### PR TITLE
Vtctld features.

### PR DESCRIPTION
**Pull request by alainjobart**
Adding features to vtctld. It's a map vtctld exports, and the ui uses to expose or not UI components.

Also reworking the menus in the dashboard, and fixing a few issues.

### CODE REVIEW

**Comment by alainjobart** [comment]
Will squash / rebuild the UI after review.

**Comment by thompsonja** [comment]
I think instead of having the various UI features disappear completely, perhaps disable/gray them out and add an explanatory tooltip, i.e. "Feature unavailable due to disable_active_reparents flag."

**Comment by alainjobart** [comment]
Talked with Josh offline about this, since the features will mostly all be enabled in open source side, and only disabled in places where they're not supported, not displaying them makes more sense.

File: web/vtctld2/src/app/dashboard/shard.component.ts [hunk 1]
```diff
@@ -72,20 +65,55 @@ export class ShardComponent implements OnInit, OnDestroy {
     });
   }
 
+  getActions() {
+    let result: MenuItem[] = [
+      {label: 'Status', items: [
+        {label: 'Validate', command: (event) => {this.openValidateShardDialog(); }},
+        {label: 'Validate Version', command: (event) => {this.openValidateVerShardDialog(); }},
```
Comment (pivanof) [inline web/vtctld2/src/app/dashboard/shard.component.ts]: I see this is copied from what was here before, but it's completely unclear what's the difference between these validates and what is validated by the first one if not version. Can you please name the first "Validate <something>"?

File: web/vtctld2/src/app/dashboard/shard.component.ts [hunk 2]
```diff
@@ -72,20 +65,55 @@ export class ShardComponent implements OnInit, OnDestroy {
...
+        {label: 'Shard Replication Positions', command: (event) => {this.openShardReplicationPosDialog(); }},
+      ]},
+      {label: 'Change', items: [
+        {label: 'Delete', command: (event) => {this.openDeleteShardDialog(); }},
+        {label: 'Tablet Externally Reparented', command: (event) => {this.openTabExtRepDialog(); }},
```
Comment (pivanof) [inline web/vtctld2/src/app/dashboard/shard.component.ts]: I think it's very confusing to have "Tablet Externally Reparented" item on a shard. Why it's not on a tablet?

File: web/vtctld2/src/app/dashboard/shard.component.ts [hunk 3]
```diff
@@ -72,20 +65,55 @@ export class ShardComponent implements OnInit, OnDestroy {
...
+      {label: 'Status', items: [
+        {label: 'Ping', command: (event) => {this.openPingTabletDialog(); }},
+        {label: 'Refresh State', command: (event) => {this.openRefreshTabletDialog(); }},
+        {label: 'Run Health Check', command: (event) => {this.openRunHealthCheckDialog(); }},
+      ]},
+      {label: 'Change', items: [
+        {label: 'Set ReadOnly', command: (event) => {this.openSetReadOnlyDialog(); }},
+        {label: 'Set ReadWrite', command: (event) => {this.openSetReadWriteDialog(); }},
+        {label: 'Start Slave', command: (event) => {this.openStartSlaveDialog(); }},
+        {label: 'Stop Slave', command: (event) => {this.openStopSlaveDialog(); }},
```
Comment (pivanof) [inline web/vtctld2/src/app/dashboard/shard.component.ts]: Should these 4 items be dependent on activeReparents as well?

File: web/vtctld2/src/app/dashboard/shard.component.ts [hunk 4]
```diff
@@ -133,9 +165,9 @@ export class ShardComponent implements OnInit, OnDestroy {
   }
 
   openDeleteShardDialog() {
-    this.dialogSettings = new DialogSettings('Delete', `Delete ${this.shardName}`,
-                                             `Are you sure you want to delete ${this.shardName}?`,
-                                             'There was a problem deleting {{shard_ref}}:');
+    this.dialogSettings = new DialogSettings('Delete', `Delete Shard ${this.keyspaceName}/${this.shardName}`,
+                                             `Are you sure you want to delete ${this.keyspaceName}/${this.shardName}?`,
```
Comment (pivanof) [inline web/vtctld2/src/app/dashboard/shard.component.ts]: I think this line should be wrapped, as well as several modified lines below.

File: web/vtctld2/src/app/dashboard/shard.component.ts [hunk 5]
```diff
@@ -164,18 +195,19 @@ export class ShardComponent implements OnInit, OnDestroy {
   }
 
   openTabExtRepDialog() {
-    this.dialogSettings = new DialogSettings('Update', `Externally Reparent a Tablet`, '',
-                                             `There was a problem there was a problem updating {{tablet_alias}}'s metatdata:`);
-    this.dialogSettings.setMessage('Changed metadata in the topology server for {{tablet_alias}}');
+    this.dialogSettings = new DialogSettings('TabletExternallyReparented',
+                                             `Run Tablet Externally Reparented in Shard ${this.keyspaceName}/${this.shardName}`, '',
+                                             `There was a problem running TER:`);
```
Comment (pivanof) [inline web/vtctld2/src/app/dashboard/shard.component.ts]: Can you avoid using the word "TER"?

File: web/vtctld2/src/app/dashboard/shard.component.ts [hunk 6]
```diff
@@ -184,35 +216,33 @@ export class ShardComponent implements OnInit, OnDestroy {
   }
 
   openEmergencyRepShardDialog() {
-    this.dialogSettings = new DialogSettings('Reparent', `Emergency Reparent Shard`, '',
-                                             'There was a problem initializing a shard master for {{shard_ref}}:');
+    this.dialogSettings = new DialogSettings('Reparent', `Emergency Reparent in Shard ${this.keyspaceName}/${this.shardName}`, '',
+                                             'There was a problem emergency reparenting shard {{shard_ref}}:');
```
Comment (pivanof) [inline web/vtctld2/src/app/dashboard/shard.component.ts]: "Emergency reparenting" doesn't seem to be in English. Better to use "problem with emergency reparenting".

File: web/vtctld2/src/app/dashboard/shard.component.ts [hunk 7]
```diff
@@ -184,35 +216,33 @@ export class ShardComponent implements OnInit, OnDestroy {
...
-    let flags = new ShardReplicationPosFlags(this.keyspaceName, this.shardName).flags;
-    this.dialogContent = new DialogContent(this.shardName, flags, {}, undefined, 'ShardReplicationPositions');
-    this.dialogSettings.toggleModal();
-  }
-
   openValidateVerShardDialog() {
-    this.dialogSettings = new DialogSettings('Validate', `Validate ${this.shardName}'s Version`, '',
-                                             'There was a problem initializing a shard master for {{shard_ref}}:');
+    this.dialogSettings = new DialogSettings('Validate', `Validate Shard ${this.keyspaceName}/${this.shardName} Versions`, '',
+                                             'There was a problem validating shard {{shard_ref}}:');
```
Comment (pivanof) [inline web/vtctld2/src/app/dashboard/shard.component.ts]: "validating shard versions"?

File: web/vtctld2/src/app/dashboard/shard.component.ts [hunk 8]
```diff
@@ -184,35 +216,33 @@ export class ShardComponent implements OnInit, OnDestroy {
...
     this.dialogSettings.setMessage(`Validated {{shard_ref}}'s Version`);
-    this.dialogSettings.onCloseFunction = this.refreshShardView.bind(this);
```
Comment (pivanof) [inline web/vtctld2/src/app/dashboard/shard.component.ts]: Why you don't need refreshShardView anymore?

File: go/vt/vtctl/reparent.go [hunk 9]
```diff
@@ -18,12 +18,15 @@ import (
 )
 
 var (
-	disableActiveReparents = flag.Bool("disable_active_reparents", false, "if set, do not allow active reparents. Use this to protect a cluster using external reparents.")
+	// DisableActiveReparents is a flag to disable active
+	// reparents for safety reasons. It is exported so the UI
+	// menus can be disabled.
```
Comment (alainjobart) [inline go/vt/vtctl/reparent.go]: Yes, the first sentence says we disable active reparents. The second sentence says it is exported so the UI can see it. I'll make it clearer.

File: go/vt/vtctld/api.go [hunk 10]
```diff
@@ -516,4 +516,30 @@ func initAPI(ctx context.Context, ts topo.Server, actions *ActionRepository, rea
...
+	// Features
+	handleAPI("features", func(w http.ResponseWriter, r *http.Request) error {
+		if err := acl.CheckAccessHTTP(r, acl.ADMIN); err != nil {
+			http.Error(w, "403 Forbidden", http.StatusForbidden)
+			return nil
+		}
+
+		resp := make(map[string]interface{})
+		if *enableRealtimeStats {
+			resp["showStatus"] = true
```
Comment (alainjobart) [inline go/vt/vtctld/api.go]: OK.

File: web/vtctld2/src/app/dashboard/shard.component.ts (hunk 1 above)
Comment (alainjobart) [inline web/vtctld2/src/app/dashboard/shard.component.ts]: I need to just remove the first Validate, it's being replaced by the topo validator. I'll leave it as is now, until it's remove.

(also I don't understand the change you're suggesting, rename the first "Validate" when it's already named Validate?).

File: web/vtctld2/src/app/dashboard/shard.component.ts (hunk 2 above)
Comment (alainjobart) [inline web/vtctld2/src/app/dashboard/shard.component.ts]: The reparent actions should also be on a tablet then. Note I also had moved it, but then moved it back, as I think all reparents should work on the shard. I think it's either:
- all reparents on the shard, with a dialog that has a dropdown with which tablet to use.
- all reparents on the master-elect tablet, with no tablet drop-down.

If you all think it should be on the tablet, I can move them, but I'll do it in a different PR.

File: web/vtctld2/src/app/dashboard/shard.component.ts (hunk 3 above)
Comment (alainjobart) [inline web/vtctld2/src/app/dashboard/shard.component.ts]: No, we do these independently of Vitess handling mastership. In fact, we even have a hook to stop slave in a way that communicates with the master process.

File: web/vtctld2/src/app/dashboard/shard.component.ts (hunk 4 above)
Comment (alainjobart) [inline web/vtctld2/src/app/dashboard/shard.component.ts]: The lint process says up to 140 characters per line. I'm just following that.

File: web/vtctld2/src/app/dashboard/shard.component.ts (hunk 5 above)
Comment (alainjobart) [inline web/vtctld2/src/app/dashboard/shard.component.ts]: Sure.

File: web/vtctld2/src/app/dashboard/shard.component.ts (hunk 6 above)
Comment (alainjobart) [inline web/vtctld2/src/app/dashboard/shard.component.ts]: Done.

File: web/vtctld2/src/app/dashboard/shard.component.ts (hunk 7 above)
Comment (alainjobart) [inline web/vtctld2/src/app/dashboard/shard.component.ts]: Done.

**Comment by alainjobart** [comment]
I'm confused on which review tool to use, so here is my feedback on the couple more comments:

- refreshShardView is only needed when a command is going to change anything displayed in the shard view. The places I removed the call from don't change anything. If you Ping a tablet, it's annoying to have the display flicker and refresh.

- using constants in that case: the only places they're used is in that go code and the corresponding JS code. So we can't really share them. Defining them as constants is a bit too much isn't it?

File: web/vtctld2/src/app/dashboard/shard.component.ts (hunk 1 above)
Comment (pivanof) [inline web/vtctld2/src/app/dashboard/shard.component.ts]: Sorry, github treated my \<something\> as html tag and didn't show it to you. So my suggestion was to add something after word "Validate", so that it was e.g. "Validate Shard" and "Validate Version" or something like that.

File: web/vtctld2/src/app/dashboard/shard.component.ts (hunk 2 above)
Comment (pivanof) [inline web/vtctld2/src/app/dashboard/shard.component.ts]: I guess I don't have a strong opinion on where these commands should be. My concern is just that a menu item saying "Tablet ... Reparented" seem to suggest that the tablet I'm selecting it on is reparented. And having such menu item on a shard is confusing. "Externally Reparent Tablet" while not ideal doesn't seem to have such strong suggestion that it should be selected on a tablet.

File: web/vtctld2/src/app/dashboard/shard.component.ts (hunk 3 above)
Comment (pivanof) [inline web/vtctld2/src/app/dashboard/shard.component.ts]: I can understand that stop/start slave can be independent from Vitess handling mastership. But how can marking master read-only or marking a replica read-write be independent from Vitess handling mastership? Or am I misunderstanding what these commands do?

File: web/vtctld2/src/app/dashboard/shard.component.ts [hunk 11]
```diff
@@ -238,47 +267,42 @@ export class ShardComponent implements OnInit, OnDestroy {
...
     let flags = new PingTabletFlags(this.selectedTablet.alias).flags;
     this.dialogContent = new DialogContent('tablet_alias', flags, {}, undefined, 'SetReadOnly');
     this.dialogSettings.toggleModal();
   }
 
   openSetReadWriteDialog() {
-    this.dialogSettings = new DialogSettings('Set', `Set ${this.selectedTablet.label} to Read/Write`, '',
+    this.dialogSettings = new DialogSettings('Set', `Set ${this.selectedTablet.label} to Read-Write`, '',
                                              `There was a problem setting ${this.selectedTablet.label} to Read/Write:`);
```
Comment (pivanof) [inline web/vtctld2/src/app/dashboard/shard.component.ts]: Did you want to change this to have dash as well (and Read Only above too)?

File: web/vtctld2/src/app/dashboard/shard.component.ts [hunk 12]
```diff
@@ -238,47 +267,42 @@ export class ShardComponent implements OnInit, OnDestroy {
...
     let flags = new PingTabletFlags(this.selectedTablet.alias).flags;
     this.dialogContent = new DialogContent('tablet_alias', flags, {}, undefined, 'SetReadWrite');
     this.dialogSettings.toggleModal();
   }
 
   openStartSlaveDialog() {
-    this.dialogSettings = new DialogSettings('Start', `Start Slave, ${this.selectedTablet.label}`, '',
+    this.dialogSettings = new DialogSettings('Start', `Start Slave on ${this.selectedTablet.label}`, '',
                                              `There was a problem starting slave, ${this.selectedTablet.label}:`);
     this.dialogSettings.setMessage(`Started Slave, ${this.selectedTablet.label}`);
```
Comment (pivanof) [inline web/vtctld2/src/app/dashboard/shard.component.ts]: Did you want to add "on" in these messages too?

File: web/vtctld2/src/app/dashboard/shard.component.ts (hunk 1 above)
Comment (alainjobart) [inline web/vtctld2/src/app/dashboard/shard.component.ts]: Done.

File: web/vtctld2/src/app/dashboard/shard.component.ts (hunk 2 above)
Comment (alainjobart) [inline web/vtctld2/src/app/dashboard/shard.component.ts]: OK changed it to 'Externally Reparent'.

File: web/vtctld2/src/app/dashboard/shard.component.ts (hunk 3 above)
Comment (alainjobart) [inline web/vtctld2/src/app/dashboard/shard.component.ts]: These commands are not enabled / disabled by the vtctld flag, so if I made them optional here, it would be inconsistent.

That being said, I don't have a problem revisiting these commands grouping. We'd have to double check they're not used during resharding though (I don't think they are). Let's do that in a different PR though.

File: web/vtctld2/src/app/dashboard/shard.component.ts (hunk 11 above)
Comment (alainjobart) [inline web/vtctld2/src/app/dashboard/shard.component.ts]: Made all of them consistent.

File: web/vtctld2/src/app/dashboard/shard.component.ts (hunk 12 above)
Comment (alainjobart) [inline web/vtctld2/src/app/dashboard/shard.component.ts]: Fixed these.

**Comment by alainjobart** [comment]
My comments are not really visible easily, here is the last one, regarding the set read-only / set read-write:

These commands are not enabled / disabled by the vtctld flag, so if I made them optional here, it would be inconsistent.

That being said, I don't have a problem revisiting these commands grouping. We'd have to double check they're not used during resharding though (I don't think they are). Let's do that in a different PR though.

@pivanof PTAL
